presto-main/src/main/java/com/facebook/presto/execution/scheduler/PartialResultQueryTaskTracker.java

```diff
@@ -0,0 +1,116 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler;
+
+import com.facebook.presto.execution.PartialResultQueryManager;
+import com.facebook.presto.execution.RemoteTask;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.TaskInfo;
+import com.facebook.presto.spi.PrestoWarning;
+import com.facebook.presto.spi.WarningCollector;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.facebook.presto.spi.StandardWarningCode.PARTIAL_RESULT_WARNING;
+import static com.google.common.collect.Sets.SetView;
+import static com.google.common.collect.Sets.difference;
+import static com.google.common.collect.Sets.newConcurrentHashSet;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class PartialResultQueryTaskTracker
+{
+    private final PartialResultQueryManager partialResultQueryManager;
+    private final double minCompletionRatioThreshold;
+    private final double timeMultiplier;
+    private final WarningCollector warningCollector;
+    private final long startTime;
+    private final Map<TaskId, RemoteTask> taskIdMap = new HashMap<>();
+    private final Set<TaskId> completedTaskIds = newConcurrentHashSet();
+    private final AtomicBoolean addedToQueryManager = new AtomicBoolean();
+
+    private long maxEndTime;
+    private boolean taskSchedulingCompleted;
+
+    public PartialResultQueryTaskTracker(
+            PartialResultQueryManager partialResultQueryManager,
+            double minCompletionRatioThreshold,
+            double timeMultiplier,
+            WarningCollector warningCollector)
+    {
+        this.partialResultQueryManager = requireNonNull(partialResultQueryManager, "partialResultQueryManager is null");
+        this.minCompletionRatioThreshold = minCompletionRatioThreshold;
+        this.timeMultiplier = timeMultiplier;
+        this.warningCollector = requireNonNull(warningCollector, "warningCollector is null");
+        this.startTime = System.nanoTime();
+    }
+
+    private double getTaskCompletionRatio()
+    {
+        return (double) completedTaskIds.size() / (double) taskIdMap.size();
+    }
+
+    public void trackTask(RemoteTask task)
+    {
+        taskIdMap.put(task.getTaskId(), task);
+    }
+
+    public void recordTaskFinish(TaskInfo taskInfo)
+    {
+        completedTaskIds.add(taskInfo.getTaskId());
+        if (taskSchedulingCompleted && !addedToQueryManager.get()) {
+            if (getTaskCompletionRatio() >= minCompletionRatioThreshold) {
+                // Completion ratio greater than equal to threshold. Query eligible for partial results
+                if (!addedToQueryManager.compareAndSet(false, true)) {
```

Comment:
I'm wondering if it is possible to combine all these if statements, for example:

`taskSchedulingCompleted && getTaskCompletionRatio() >= minCompletionRatioThreshold && addedToQueryManager.compareAndSet(false, true)`

Then maybe we could just extract this to a function `shouldAddToQueryManager`.